[Spring 지하철 노선도 - 3단계] 알렉스(이영환) 미션 제출합니다.

yxxnghwan opened the pull request:
안녕하세요 미르! 저번단계 PR 이후로 시간이 꽤 흘러서야 새 PR을 제출하네요...😥

이번 단계도 잘 부탁드립니다!💪

저번 주신 피드백에 대해서 궁금한 점 남깁니다!
> RuntimeException을 상속받는 CustomException을 만든 후, CustomException을 활용한다면 ExceptionHandler를 줄일 수 있을 것 같아요.
> 
> RuntimeExceptoin -> CustomException (HttpStatus, Message) -> DuplicateNameException
>                                                           -> NotFoundStationException (NoSuchElementException 대신 사용)

미르가 주신 피드백 참고하여 해당 서비스에서 다루는 커스텀 예외들을 묶어서 계층 구조로 관리하게 했고 각각 HttpStatus를 가지게 해서,
최상단에 있는 커스텀 예외만 ExceptionAdvice로 핸들링하게 구조를 변경해봤는데 미르가 의도한 방향이 맞는지 모르겠네요! 확인 부탁드려요!

> 저는 오히려 Service Layer이기 때문에 DuplicateKeyException를 사용해도 괜찮다고 생각해요.
> 저는 Service Layer가 Domain이 오염되지 않도록 캡슐화하여 Domain 순서 보장을 해주고 기술적인 내용(Transaction 등)을 처리하는 역할이라고 생각하기 때문이예요.
> 
> (JPA를 사용할 때는 DataIntegrityViolationException를 사용해야합니다. H2도 같은 에러를 발생하기 때문에 DataIntegrityViolationException를 추천드려요.)
> 
> try {
>     final LineEntity lineEntity = new LineEntity(Line.createWithoutId(name, color));
>     final LineEntity savedLineEntity = lineDao.save(lineEntity);
>     return new Line(savedLineEntity.getId(), savedLineEntity.getName(), savedLineEntity.getColor());
> } catch (DataIntegrityViolationException e) {
>     throw new DuplicateNameException("[ERROR] 이미 존재하는 노선 이름입니다.");
> }
> 와 같이 작성한다면 에러 메시지를 친절하게 작성할 수 있지 않을까요?

생각해보니 이미 서비스레이어에서 `@Transactional`이라는 어노테이션을 달고 있기 때문에 DB를 참조하는 상태였네요!
미르의 추천대로 `DataIntegrityViolationException`을 서비스에서 캐치하는 방식으로 변경해봤는데요,
우려되는 점이 저는 노선의 color를 유니크 키로 잡지 않았지만, 유니크 키로 잡은 크루들도 있더라구요.
이 경우엔 `DataIntegrityViolationException`이 발생해도 어떤 컬럼에서 문제가 생겨 발생했는지 캐치하기 어려울 것 같은데,
그럴땐 처음 했던 방식대로 findByName, findByColor등을 사용해서 중복을 막는 방법밖에 없을까요..?

yxxnghwan committed:
docs: 리드미에 구현할 기능 목록 작성

yxxnghwan committed:
feat: 찾는 리소스가 존재하지 않을 경우 404 not found를 응답하도록 변경

yxxnghwan committed:
refactor: sql 예약어 대문자로 변경

yxxnghwan committed:
fix: 업데이트시 이름 중복 검증시 같은 이름으로 조회할 때 변경하려는 데이터 떄문에 무조건 true로 판단되던 현상 수정

yxxnghwan committed:
refactor: 이름 중복 로직 검증시 따로 검증 쿼리를 실행하지 않고 DataIntegrityViolationException을 캐치하여 검증

yxxnghwan committed:
refactor: 커스텀 예외 구체화, 계층 정리

yxxnghwan committed:
feat: 섹션 추가 기능 구현, 라인 추가 기능 요구사항에 맞게 변경

yxxnghwan committed:
test: 테스트 중복 로직 fixture로 분리

yxxnghwan committed:
docs: 구현 기능 목록 체크

yxxnghwan committed:
feat(domain): 구간 추가 기능 구현

yxxnghwan committed:
fix(domain): 구간 추가 기능 구현로직 조건 추가

yxxnghwan committed:
test(domain): 구간 추가 기능 테스트 케이스 추가

yxxnghwan committed:
docs: 기능 구현 목록 체크

yxxnghwan committed:
feat: 역 사이에 새로운 역을 등록할 경우 기존 역 사이 길이보다 크거나 같으면 예외 발생

yxxnghwan committed:
test(dao): 구간 저장 기능 테스트코드 작성

yxxnghwan committed:
feat(service): 구간 등록 기능 구현

yxxnghwan committed:
fix(service): 구간 저장 기능 수정

yxxnghwan committed:
feat: 구간 추가 API 구현

yxxnghwan committed:
feat(domain): 구간 리스트에서 특정 역 삭제 기능 구현

yxxnghwan committed:
refactor(domain): 노선에서 역 삭제 기능 메서드 분리

yxxnghwan committed:
fix(Sections): 정렬, 중간삭제 기능 버그 수정

yxxnghwan committed:
feat(service): 구간 삭제 기능 구현

yxxnghwan committed:
feat(service): 구간삭제 테스트 케이스 추가

yxxnghwan committed:
feat(controller): 구간제거 API 구현

yxxnghwan committed:
feat(controller): LineResponseDto 형식에 맞게 수정

yxxnghwan committed:
test: 컴파일 에러 수정

yxxnghwan committed:
chore: 불필요한 import문 및 공백 제거

yxxnghwan committed:
feat: 노선의 색깔 데이터 길이가 20 이상이면 예외발생 기능 구현

yxxnghwan committed:
docs(README): 기능 구현 목록 체크

yxxnghwan committed:
feat: 커스텀 예외 추가

yxxnghwan committed:
feat: `@Transactional`어노테이션 적용

yxxnghwan committed:
refactor(SectionService): resisterFirst() 메서드 반환형 void로 변경

yxxnghwan committed:
refactor(LineService): 메서드 분리

yxxnghwan committed:
chore: 불필요한 import, 공백, 메서드 삭제

yxxnghwan committed:
test: 테스트 케이스 추가

yxxnghwan committed:
refactor(dao): findById에서 Optional.ofNullable()을 사용해서 반환하도록 변경

yxxnghwan committed:
test(dao): 테스트 케이스 추가

yxxnghwan committed:
feat(sql): 섹션 테이블 외래키 설정

yxxnghwan committed:
refactor: Sections 최소 사이즈 상수로 분리

yxxnghwan committed:
style: 메서드 파라미터 개행

ddu0422 commented on src/main/java/wooteco/subway/controller/ExceptionAdvice.java:
```diff
@@ -1,27 +1,20 @@
 
-    @ExceptionHandler(DuplicateNameException.class)
-    public ResponseEntity<String> illegalArgument(final Exception exception) {
+    @ExceptionHandler(SubwayException.class)
```
잘 이해하셨네요 ! 👍🏻 👍🏻 
만약, Custom 에러가 추가된다면 해당 파일을 고칠 필요가 없는 형태가 되었네요 😄

ddu0422 commented on src/main/java/wooteco/subway/domain/Line.java:
```diff
@@ -1,19 +1,31 @@
+    }
+
+    private void validateColor(final String color) {
+        if (color.length() > 20) {
```
상수로 선언하는건 어떨까요?

yxxnghwan replied:
깜빡했네요! 상수로 관리하게 했습니다!

ddu0422 commented on src/main/java/wooteco/subway/domain/Line.java:
```diff
@@ -1,19 +1,31 @@
+
+    private void validateColor(final String color) {
+        if (color.length() > 20) {
+            throw new LineColorLengthException("[ERROR] 노선 색은 20자 이하여야 합니다.");
```
에러 클래스에서 에러 메시지를 관리하면 어떨까요?

yxxnghwan replied:
해당 에러는 저 상황에서만 사용하는 에러라서 에러클래스에서 메시지를 관리해도 별 문제 없을 것 같네요. 일단은 에러 클래스에서 관리하도록 변경했습니다.

만약, 여러 상황에서 발생할 수 있는 커스텀 에러를 정의한다면 어떤 상황에서 발생했는 지 친절한 에러 메시지를 작성하기 위해 메시지를 파라미터로 받는게 좋을 것 같은데요, 이런 경우에 에러 클래스를 정의하는 일관성을 위해 모두 메시지를 받는 형식으로 통일하는 식으로 하다 보니 저렇게 작성하게 되었던 것 같아요.
그리고 만약 저 에러가 여러 군데에서 발생할 수 있게 된다면 생성자의 시그니처를 변경해야할 것 같은데... 
미르는 필요한 경우에만 메시지를 파라미터로 받는걸 선호하시나요?

ddu0422 commented on src/main/java/wooteco/subway/domain/Sections.java:
```diff
@@ -0,0 +1,292 @@
+                        section.getDownStation().equals(station)
+                                || section.getUpStation().equals(station)
+                );
+    }
```
Section 클래스에게 메시지를 보내는 건 어떨까요?

yxxnghwan replied:
Sections에서 너무 모든 값을 꺼내서 로직을 작성하면서 객체지향적이지 못하게 코드를 짜고 있었네요...😂
Station과 Section에 책임을 옮겨봤습니다!

ddu0422 commented on src/main/java/wooteco/subway/domain/Sections.java:
```diff
@@ -0,0 +1,292 @@
+                .orElseThrow(() -> new NotFoundStationException("[ERROR] 해당 구간이 존재하지 않습니다."));
+    }
+
+    private boolean isOnlyUpStation(final LinkedList<Section> sections, final Station station) {
+        return sections.stream()
+                .anyMatch(section -> section.getUpStation().equals(station))
+                && sections.stream()
+                .noneMatch(section -> section.getDownStation().equals(station));
+    }
```
UpStation과 DownStation이 동일한 경우는 어떻게 될까요?

yxxnghwan replied:
Section 클래스를 만들 때 같은 역으로 구간을 등록하지 못하도록 검증하는 로직을 추가했습니다!
같은 Section 클래스가 만들어질 수 없기 때문에 해당 로직에서 UpStation과 DownStation이 동일한 경우는 배제하고 작성해도 될 것 같아요!

ddu0422 commented on src/main/java/wooteco/subway/service/LineService.java:
```diff
@@ -4,69 +4,121 @@
+                                         final int distance) {
+        final LineEntity savedLineEntity = lineDao.save(LineEntity.createWithoutId(name, color));
+        final Station upStation = stationService.searchById(upStationId);
+        final Station downStation = stationService.searchById(downStationId);
```
saveLineAndFirstSection 메서드가 한가지 일에 집중할 수 있도록 register메서드에서 Station을 미리 찾고 넘겨주는 방식은 어떤가요?

yxxnghwan replied:
resister 메서드에서 Station을 미리 찾고 saveLineAndFirstSection메서드로 넘겨주는 방식으로 변경했습니다!

ddu0422 commented on src/test/java/wooteco/subway/acceptance/LineAcceptanceTest.java:
```diff
@@ -156,4 +185,82 @@ void removeLine() {
+                .contentType(MediaType.APPLICATION_JSON_VALUE)
+                .post("/lines/" + createdLine.getId() + "/sections")
+                .then().log().all()
+                .extract();
```
서버에 요청하는 내용이 Fixture에서 다른 생성 메서드와 상당히 비슷한 형태이지 않을까요?

yxxnghwan replied:
구간 생성 요청도 다른 생성 요청들 처럼 픽스쳐 메서드로 분리했습니다!

ddu0422 commented on src/test/java/wooteco/subway/domain/SectionsTest.java:
```diff
@@ -0,0 +1,250 @@
+    void emptyListSections() {
+        assertThatThrownBy(() -> new Sections(new LinkedList<>()))
+                .isInstanceOf(NoElementSectionsException.class)
+                .hasMessage("[ERROR] 최소 한 개의 구간이 있어야 Sections 객체를 생성할 수 있습니다.");
```
에러 메시지가 변경이 된다면, 테스트가 깨지지 않을까요?
만약, 동일한 에러 메시지를 전부 수정해주는 작업이 필요하다면 유지보수 비용이 커지지 않을까요?

yxxnghwan replied:
TDD 관점에서 에러 메시지를 변경해야 한다면 순서가

1. 테스트 코드에서 에러 메시지 수정
2. 테스트 실패
3. 프로덕션 코드 수정
4. 테스트 통과

이게 맞다고 판단되어서 일부러 저렇게 했는데요,
에러 메시지를 상수화하여 관리하고 테스트코드에서 사용한다면 에러 메시지 상수의 값을 변경해도 테스트 코드를 수정할 필요는 없겠지만, 상수를 그대로 가져다 비교하는건 테스트 코드를 짜는 의미가 없는 것 같다고 생각돼서 저렇게 작성했습니다!

미르 의견도 알려주세요!

ddu0422 replied:
저도 처음에 알렉스와 같은 의견으로 메시지를 일일히 작성했었어요. 하지만 현재는 생각이 많이 바뀌었답니다 😄

저는 `노선 색은 20자 이하여야 합니다.` 가 `노선색은 20자이하로 입력해주세요.` 로 변경되는 것 보단, 20자라는 비즈니스 규칙이 중요하다고 생각해서 20이라는 상수를 받아서 해당 값 정도만 확인해요.
메시지도 정말 중요한 부분이지만, 최소한의 비용으로 최대한의 효과를 내려면 유지보수 비용을 줄여야한다고 생각하기 때문이예요 !

사용자에게 의미있는 메시지 전달은 중요하지만, 비즈니스 규칙보다는 중요도가 낮다고 생각하기 때문이예요 !

알렉스는 LineColorLengthException 에러 메시지 중에서 가장 중요한 부분이 어떤 부분이라고 생각하나요?

https://github.com/woowacourse/atdd-subway-map/blob/8d72c10dbd1e4fa3fafaa33e78aa3e9e9be19a51/src/main/java/wooteco/subway/exception/LineColorLengthException.java#L7-L9

ddu0422 commented on src/test/java/wooteco/subway/service/SectionServiceTest.java:
```diff
@@ -0,0 +1,163 @@
+    @DisplayName("노선 id, 상행역 id, 하행역 id, 길이를 입력받아 첫 구간을 등록한다.")
+    @Test
+    void resisterFirst() {
+        final Long upStationId = Fixture.saveStation("선릉역");
+        final Long downStationId = Fixture.saveStation("잠실역");
```
BeforeEach를 활용하거나 Nested Class로 선언하여 중복을 제거해보는 건 어떨까요?

yxxnghwan replied:
구간을 생성하려면 최소 두 개의 역이 필요해서 해당 테스트 클래스에서 모든 메서드에 역 두개를 생성하는 로직이 필요하네요!

BeforeEach를 활용해서 중복을 제거해봤습니다

ddu0422 commented on src/main/java/wooteco/subway/domain/Sections.java:
```diff
@@ -0,0 +1,292 @@
+        final Section frontSection = Section.createWithoutId(
+                oldSection.getUpStation(),
+                newUpStation,
+                oldSection.getDistance() - distance);
```
Distance도 하나의 도메인 규칙을 가진다고 생각해서 저라면 Distance 객체를 만든 후 해당 객체에서 처리할 것 같은데 알렉스의 생각은 어떤가요?

yxxnghwan replied:
이번 미션에서 너무 DB와 스프링만 신경쓰다 보니 값객체를 사용하는걸 고려하지 않고 해버렸네요 ㅠㅠ
Distance 뿐만 아니라 Name, LineColor 라는 값 객체를 정의했습니다!

ddu0422 replied:
subtract 메서드등을 만들어 Distance에게 책임을 부여할 수 있지 않을까요?

ddu0422 commented on src/main/java/wooteco/subway/service/LineService.java:
```diff
@@ -4,69 +4,121 @@
+                         final int distance) {
+        try {
+            return saveLineAndFirstSection(name, color, upStationId, downStationId, distance);
+        } catch (DataIntegrityViolationException e) {
```
> 생각해보니 이미 서비스레이어에서 @Transactional이라는 어노테이션을 달고 있기 때문에 DB를 참조하는 상태였네요!
미르의 추천대로 DataIntegrityViolationException을 서비스에서 캐치하는 방식으로 변경해봤는데요,
우려되는 점이 저는 노선의 color를 유니크 키로 잡지 않았지만, 유니크 키로 잡은 크루들도 있더라구요.
이 경우엔 DataIntegrityViolationException이 발생해도 어떤 컬럼에서 문제가 생겨 발생했는지 캐치하기 어려울 것 같은데,
그럴땐 처음 했던 방식대로 findByName, findByColor등을 사용해서 중복을 막는 방법밖에 없을까요..?

우선 Color의 중복을 막고 싶다면 우선 유니크 키로 설정하는 방향이 가장 좋다고 생각해요.
데이터 중복을 막는 가장 간단한 방법이기 때문이죠.

또한, findByName, findByColor등을 사용해서 중복을 막는 방법도 상당히 괜찮다고 생각하기 때문에 어떠한 방법을 사용하셔도 상관 없을 듯 합니다 😄 
언제나 안전하고 이해하기 쉬운 코드를 작성하는 방향으로 개발만 한다면 어떤 방식을 사용해도 괜찮다고 생각합니다 !

ddu0422 reviewed (changes requested):
안녕하세요 알렉스 ! 미르예요 :)

3단계도 잘 구현해주셨네요 ㅎ_ㅎ
간단한 코멘트 남겨드렸으니 확인해주시면 감사하겠습니다 :)

마지막까지 화이팅해보아요 !_!

yxxnghwan committed:
refactor: LineColor 값객체로 설계, 상수 분리

yxxnghwan committed:
refactor: LineColorLengthException 예외 메시지를 예외 클래스에서 관리하도록 변경

LineColorLengthException이 발생하는 곳은 LineColor 한 곳이므로 LineColorLengthException 예외 메시지를 예외 클래스에서 관리하도록 변경

yxxnghwan committed:
refactor: Sections.isResistedStation() 메서드의 판단로직을 Section으로 옮김

yxxnghwan committed:
refactor(LineService): saveLineAndFirstSection() 메서드 시그니처 변경

yxxnghwan committed:
refactor: Distance 값 객체 정의

yxxnghwan committed:
rename: 클래스 이름 오타 수정

yxxnghwan committed:
refactor: Name 값 객체 정의

yxxnghwan committed:
test(SectionServiceTest): 초기 구간 생성을 위한 station 두 개 BeforeEach에서 생성

yxxnghwan committed:
test: 구간 생성 요청, 구간 삭제 요청 픽스쳐 메서드로 분리

yxxnghwan committed:
test: 라인 생성 요청 파라미터 픽스쳐 분리

yxxnghwan committed:
test: 테스트 케이스 추가

yxxnghwan committed:
chore: 불필요한 공백 및 임포트문 제거

yxxnghwan committed:
refactor: 역이 구간에 등록되어있는 지 확인하는 로직을 Section 클래스로 이동

yxxnghwan committed:
feat: 같은 역으로 구간등록하는 케이스 검증

yxxnghwan commented:
안녕하세요 미르! 피드백 반영해서 새로 리뷰요청 드립니다!

몇 가지 질문을 기존 코멘트에 남겼습니다! 잘 부탁드려요!

ddu0422 commented on src/main/java/wooteco/subway/domain/Distance.java:
```diff
@@ -0,0 +1,44 @@
+            throw new MinimumSectionDistanceException();
+        }
+    }
+
+
```
공백 한번 확인해주세요!

ddu0422 commented on src/test/java/wooteco/subway/domain/StationTest.java:
```diff
@@ -0,0 +1,44 @@
+    @DisplayName("구간 리스트에 해당 지하철역이 등록되어 있는 지 확인한다. - 참")
+    @Test
+    void isResistedStationTrue() {
+        final Station station1 = new Station(1L, "선릉역");
+        final Station station2 = new Station(2L, "잠실역");
+        final Station station3 = new Station(3L, "강남역");
+        final Station station4 = new Station(4L, "역삼역");
+        final List<Section> sections = List.of(
+                new Section(1L, station1, station2, 10),
+                new Section(2L, station3, station1, 10),
+                new Section(3L, station3, station4, 5)
+        );
```
이런 부분도 BeforeEach로 리팩토링할 수 있지 않을까요?
TDD Cycle을 한번 확인해보는건 어떨까요?

ddu0422 commented on src/test/java/wooteco/subway/mockDao/MockLineDao.java:
```diff
@@ -2,49 +2,72 @@
 
+    private boolean existName(final String name) {
+        final Optional<LineEntity> any = store.values().stream()
+                .filter(lineEntity -> lineEntity.getName().equals(name))
```
메시지를 보내는 건 어떨까요?

ddu0422 commented on src/test/java/wooteco/subway/mockDao/MockLineDao.java:
```diff
@@ -2,49 +2,72 @@
     }
 
     public LineEntity save(final LineEntity lineEntity) {
+        if (existName(lineEntity.getName())) {
+            throw new DataIntegrityViolationException("이름 중복되어서 라인 추가 불가능");
+        }
```
[Optional Best Practice](https://ducmanhphan.github.io/2019-12-06-Best-practice-for-Optional-in-Java/)를 확인하여 다른 방법으로 풀어보면 어떨까요?

ddu0422 reviewed (approved):
안녕하세요 알렉스 ! 미르예요 :)

구현 정말 잘해주셨어요 :) 충분히 머지해도 될 것 같아서 머지할게요 !
코멘트 남긴 부분 관련하여 궁금한 점이 있으면 코멘트 남겨주시거나 DM 주세요 !

이번 미션도 고생하셨습니다 ㅎ_ㅎ 남은 미션도 화이팅이예요 !!
